django/test/testcases.py

```diff
@@ -150,6 +150,48 @@ def __call__(self):
         )
 
 
+class _AssertSignalSentContext(object):
+    def __init__(self, test_case, signal, required_kwargs=None):
+        self.test_case = test_case
+        self.signal = signal
+        self.required_kwargs = required_kwargs
+
+    def _listener(self, sender, **kwargs):
+        self.signal_sent = True
+        self.received_kwargs = kwargs
+        self.sender = sender
+
+    def __enter__(self):
+        self.signal_sent = False
+        self.received_kwargs = {}
+        self.sender = None
+        self.signal.connect(self._listener)
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.signal.disconnect(self._listener)
+        if not self.signal_sent:
+            self.test_case.fail('Signal was not sent.')
```

Comment:
@timgraham unfortunately this cannot be achieved unless we pass a `name` argument on `Signal` instantiation as class objects don't keep a reference of where they were defined. It was already proposed by @jarshwah in his first comment.